srsenb/src/upper/rrc.cc

```diff
@@ -1881,6 +1901,113 @@ void rrc::ue::send_ue_cap_enquiry()
 
 /********************** HELPERS ***************************/
 
+bool rrc::ue::select_security_algorithms() {
+  // Each position in the bitmap represents an encryption algorithm:
+  // “all bits equal to 0” – UE supports no other algorithm than EEA0,
+  // “first bit” – 128-EEA1,
+  // “second bit” – 128-EEA2,
+  // “third bit” – 128-EEA3,
+  // other bits reserved for future use. Value ‘1’ indicates support and value
+  // ‘0’ indicates no support of the algorithm.
+  // Algorithms are defined in TS 33.401 [15].
+  // Note: information missing
+
+  bool enc_algo_found = false;
+  bool integ_algo_found = false;
+  bool zero_vector = true;
+  int i = 0;
+  for (i = 0; i < srslte::CIPHERING_ALGORITHM_ID_N_ITEMS; i++) {
+    switch (parent->cfg.eea_preference_list[i]) {
+    case srslte::CIPHERING_ALGORITHM_ID_EEA0:
+      // “all bits equal to 0” – UE supports no other algorithm than EEA0,
+      zero_vector = true;
+      for (int j; j < LIBLTE_S1AP_ENCRYPTIONALGORITHMS_BIT_STRING_LEN; j++) {
+        if (security_capabilities.encryptionAlgorithms.buffer[j]) {
+          zero_vector = false;
+        }
+      }
+      if (zero_vector == true) {
+        cipher_algo = srslte::CIPHERING_ALGORITHM_ID_EEA0;
+        enc_algo_found = true;
+        break;
+      }
+      break;
+    case srslte::CIPHERING_ALGORITHM_ID_128_EEA1:
+      // “first bit” – 128-EEA1,
+      if (security_capabilities.encryptionAlgorithms
+              .buffer[srslte::CIPHERING_ALGORITHM_ID_128_EEA1 - 1]) {
+        cipher_algo = srslte::CIPHERING_ALGORITHM_ID_128_EEA1;
+        enc_algo_found = true;
+        break;
+      }
+      break;
+    case srslte::CIPHERING_ALGORITHM_ID_128_EEA2:
+      // “second bit” – 128-EEA2,
+      if (security_capabilities.encryptionAlgorithms
+              .buffer[srslte::CIPHERING_ALGORITHM_ID_128_EEA2 - 1]) {
+        cipher_algo = srslte::CIPHERING_ALGORITHM_ID_128_EEA2;
+        enc_algo_found = true;
+        break;
+      }
+      break;
+    default:
+      enc_algo_found = false;
+      break;
+    }
+    if (enc_algo_found == true) {
+      break;
+    }
+  }
+
+  for (i = 0; i < srslte::INTEGRITY_ALGORITHM_ID_N_ITEMS; i++) {
+    switch (parent->cfg.eia_preference_list[i]) {
+    case srslte::INTEGRITY_ALGORITHM_ID_EIA0:
+      // “all bits equal to 0” – UE supports no other algorithm than EEA0,
+      zero_vector = true;
+      for (int j; j < LIBLTE_S1AP_INTEGRITYPROTECTIONALGORITHMS_BIT_STRING_LEN;
```

Comment:
This j also seems uninitialized

Reply:
Yes, setting both to zero breaks it ;-)

Reply:
Ok, I'm going to investigate. Not initializing seems wrong to me, no sure why setting it to zero is breaking it.